### ISSUE TITLE
PyArray_DescrCheck doesn't work anymore if the consumer library was compiled with an older NumPy version

### ISSUE THREAD

**Issue by xhochy** [numpy/numpy#17913]
Libraries using numpy's C-API should always build with an older release of numpy as the version of numpy you have built with is the oldest version you can support but you can support all newer versions. Thus conda-forge is currently using `numpy=1.16` as the version to build packages with for most parts of their supported matrix: https://github.com/conda-forge/conda-forge-pinning-feedstock/blob/c3c4f69b09b9a179ac304fefcf9398f2e9764463/recipe/conda_build_config.yaml#L555-L560 At the time of writing, this means that `1.16.5`(!) is used in builds as `1.16.6` hasn't been backported.

The version of `numpy` used in building also defines what implementations of the C-macros are baked in. For the referenced `PyArray_DescrCheck`, this means that implementations/builds using pre-1.16.6 don't include https://github.com/numpy/numpy/commit/78070f07b316951b4fd278312aa79a4a7a1ced4b Thus all packages built on conda-forge that use `PyArray_DescrCheck` and the wheels on PyPI / packages in distributions that were built with an older NumPy version _will break as in the upcoming 1.20 release dtypes will be actual subclasses_.

A good example here is to use the `pyarrow` packages built on conda-forge and upgrade to the latest master. Using the `pyarrow.array(np.array([1, 2])` you can trigger that error:

```
pyarrow.array(np.array([1, 2]))
---------------------------------------------------------------------------
ArrowTypeError                            Traceback (most recent call last)
<ipython-input-19-ea54df46d052> in <module>
----> 1 pyarrow.array(np.array([1, 2]))

~/miniconda3/envs/kartothek/lib/python3.7/site-packages/pyarrow/array.pxi in pyarrow.lib.array()

~/miniconda3/envs/kartothek/lib/python3.7/site-packages/pyarrow/array.pxi in pyarrow.lib._ndarray_to_array()

~/miniconda3/envs/kartothek/lib/python3.7/site-packages/pyarrow/array.pxi in pyarrow.lib._ndarray_to_type()

~/miniconda3/envs/kartothek/lib/python3.7/site-packages/pyarrow/error.pxi in pyarrow.lib.check_status()

ArrowTypeError: Did not pass numpy.dtype object
```

The issue here is fixed technically in all relevant releases in NumPy and thus from a pure code perspective, this issue can be closed. From an ecosystem perspective this though will lead to quite some breakage.

On the conda-forge side, I will propose to rebuild all latest packages using `numpy=1.16.6` so that they will work again but this won't fix old installations and especially existing wheels on PyPI.

**Comment by eric-wieser** [comment]
If we don't already have a release note about this, we should make sure that we add one.

**Comment by xhochy** [comment]
It's in the 1.16.6 changelog and part of NEP 41 at least.

**Comment by eric-wieser** [comment]
Right, but it sounds like 1.20 is the first release that actually breaks the interface fixed in 1.16.x?

**Comment by xhochy** [comment]
Yes, this popped up last night in CI as `1.20.0rc1` was released to PyPI and the jobs using `--pre` to install `numpy` and friends were failing everywhere suddenly.

**Comment by charris** [comment]
@xhochy Can you confirm that building against 1.16.6 fixes the problem? I don't think we have a good way to handle the breakage as the old macro doesn't seem to work. We could change the ABI version in `numpy/core/setup_common.py`, but that would force everyone to compile against 1.20 and that is undesireable. What about oldest-supported-numpy?

**Comment by charris** [comment]
The oldest-supported-numpy mechanism is platform/python dependent, so doesn't seem to have any relevance here.

**Comment by charris** [comment]
@rgommers Ping.

**Comment by rgommers** [comment]
@charris the main issue here is already released wheels that will break with a 1.20.0 release. If there are too many, we may need to do something - a workaround, a revert, make this NumPy 2.0, or whatever. I suspect the problem is not that widespread, but if it is we can't just proceed as planned.

`oldest-supported-numpy` isn't widely used yet, it's just taking off. Probably little relevance; it's the runtime dependencies here that need fixing first.

**Comment by rgommers** [comment]
Weird coincidence, I was just picking up on https://github.com/scipy/scipy/pull/12862 (upper bounds on build-time and run-time dependencies). We may need something like that as a structural solution now that we have a real use case.




> We could change the ABI version in `numpy/core/setup_common.py`, but that would force everyone to compile against 1.20 and that is undesireable.

The NumPy ABI version isn't reflected in wheel name or metadata right? So by itself it won't fix things, just means one gets a different error at runtime?

**Comment by seberg** [comment]
The simple solution is to just not use the macro and use `PyTyoe_Check` instead.  Repeating the release note in the 1.20 notes again sounds good to me, though.

**Comment by rgommers** [comment]
@seberg that does not prevent breaking already released wheels. 

I assume the change in 1.20 is a ton of work that's hard to revert or put a workaround for in the 1.20 release?

**Comment by seberg** [comment]
@rgommers its one that I announced very long ago that is is 100% unavoidable, so yes, as already said above I mentioned it in NEP 41.
We broke forward compatible (This issue is C, but it affects Python just as much, SciPy had the same issue in Python, I just fixed it long ago).  Forward compatible always breaks eventually?

The only (unfortunate) difference for with this change is that we can't put a DeprecationWarning on...

**Comment by rgommers** [comment]
The PR description in gh-14605 is useful:
_This becomes a problem if we allow subclassing `PyArrayDescr_Type`, and should be backported to any version we still support to prevent issues when using a newer NumPy with code compiled with an older NumPy. Searching scipy/pandas/astropy/numba only turned up a use inside numba. Once this is merged we should submit an issue there to work around this for old versions._ 

It looks like that Numba issue wasn't actually filed, and they encountered the failure in July: https://github.com/numba/numba/issues/6041

In that same PR @seberg says: _Note that in the unlikely event that this becomes an ABI breaking issue, we might have to backport even to very old versions._

That doesn't actually fix ABI breakage ..... you'd have to wait for quite a while till the affected packages have new versions out - built against a NumPy version including the fix - and users are fairly likely to have upgraded to those new versions.

Also, when for example PyArrow would build against `1.16.6`, that would also mean requiring the very latest bugfix releases of 1.17.x and 1.18.x. And if I read the 1.17.4/5 release notes right, there is no 1.17.x release which includes this fix?

**Comment by charris** [comment]
@rgommers It is in 17.3. Searching the changelog shows it.
```
charris@fc [numpy.git (maintenance/1.20.x)]$ grep -r 'properly define PyArray_Descr' doc/changelog/
doc/changelog/1.16.6-changelog.rst:* `#14686 <https://github.com/numpy/numpy/pull/14686>`__: BUG: properly define PyArray_DescrCheck
doc/changelog/1.18.0-changelog.rst:* `#14605 <https://github.com/numpy/numpy/pull/14605>`__: BUG: properly define PyArray_DescrCheck{,Exact}
doc/changelog/1.17.3-changelog.rst:* `#14687 <https://github.com/numpy/numpy/pull/14687>`__: BUG: properly define PyArray_DescrCheck
```

**Comment by seberg** [comment]
@rgommers the point is that any affected downstream project can trivially fix their code to be compatible with all NumPy versions. The macro is litertally:
```
#define PyArray_DescrCheck(op) PyObject_TypeCheck(op, &PyArrayDescr_Type)
```
It is identical to Python code using `type(obj) is np.dtype` instead of `isinstance(obj, np.dtype)`.  Which requires the exact same code change (except in Python and not in C).

Sorry for forgetting to ping Numba (probably), but I am not sure where you are going... You seem to consider which version to compile against, but since the code change required is trivial, that isn't an actual issue?

Lets discuss how to nicely announce how to fix it in the 1.20 release notes, or someone do a BigQuery if you have access to actually ping projects directly, but discussing what to do about the change itself will quite frankly only tire me.

**Comment by rgommers** [comment]
> Forward compatible always breaks eventually?

Well yes, but if your NEP had said it would break PyArrow and Numba, we'd probably have made a different plan - like consider making this a NumPy 2.0.

Also, the backport plan needs reconsidering - you cannot sensibly backport ABI-changing fixes beyond one minor release I think, that may cause as many problems as it solves.

**Comment by rgommers** [comment]
> Sorry for forgetting to ping Numba (probably), but I am not sure where you are going... You seem to consider which version to compile against, but since the code change required is trivial, that isn't an actual issue?
>
> ...but discussing what to do about the change itself will quite frankly only tire me.

You are completely ignoring what I'm saying about _already released wheels_. I'm not jumping to conclusions here, but I'd like to know the actual impact and then make a real decision rather than pre-emptively say "not our problem". 

Conda-forge can (and likely will) rebuild all affected packages. On PyPI it's possible too, with `.postX` wheels, but harder to coordinate. If it's only PyArrow that should be doable though.

**Comment by seberg** [comment]
@rgommers I am not ignoring it, I am not understanding your point.  Yes, you have to release a new version which is compatible with NumPy 1.20.  Just like your limit on the upper supported NumPy version would do.  I do not see that *binary wheels* go into the equation.  I.e. if a project used `type(dtype) is np.dtype`, they have the same issue and also have to do a new release fixing the issue.

I am not preemptively saying "its not our problem", I have announced this issue long ago and brought it up *many many* times.  But the truth is: If we try to fix this in NumPy, you just have to revert *all* of my dtype work, and hope that you can wrangle some of it into a modified, much weirder way.  So yes, I give up!

The wheels only come into the equation when you want to start patching *old* releases of these projects (which you can, just like you can patch their Python code).

**Comment by rgommers** [comment]
> Yes, you have to release a new version which is compatible with NumPy 1.20. Just like your limit on the upper supported NumPy version would do. I do not see that _binary wheels_ go into the equation.

There's a big difference there. What I had in mind with those version constraints is making all of the following not break because of incompatible versions:

1. `pip install numpy pyarrow` in a fresh env
2. `pip install --upgrade numpy` in an env where pyarrow is already installed
3. `pip install pyarrow=='2.0.0'` in any env

If we release 1.20.0 in the current state today all of those will break. 

To address that issue:
- If PyArrow releases a new version with a fix, then (1) is fixed, and (2) and (3) will still be broken.
- PyArrow can actually make (3) work by going back and making updates to old versions by rebuilding all wheels against numpy 1.16.6 and uploading them to PyPI as `2.0.0.post0` etc. Painful but doable if they'd really wanted to.
- (2) cannot be fixed, users will run into the traceback reported in this issue, then figure out they need to upgrade

It looks like there's no good way for us to avoid any of the above. The only thing we can do is signal this more clearly in the release notes. If it turns out there's other packages affected, we could decide to consider this a significant ABI break and hence make it NumPy 2.0 to signal that. 

I hope that makes it clearer why wheels are important.

**Comment by seberg** [comment]
> I hope that makes it clearer why wheels are important.

Probably I am daft, because it doesn't :), but at this point it doesn't matter.  If PyArrow had not used this macro, but had `type(dtype) is np.dtype` in Python, the situation would be identical! So the "binary" part, I don't get.

I am not sure how PIP asserts maximum version requirements, so I am not sure how much your upper version limit will fix this issue, and trust you that it doesn't fully fix it.

> consider this a significant ABI break

Yeah, we can, except that we have to recompile the world (since the `import_array` macro enforces this strictly).  In any case, it is a breaking *API* change (it is a faulty header, you can very much compile with a NumPy 1.14.x by simply avoiding that one macro).

The only thing we could do in NumPy (and I have suggested/brought it up a few times!), is to make backport releases years back to fix the header also in `1.14.x`, etc. but to have a real effect you would probably have had to do that a year ago, and nobody was interested in doing such a weird release at the time considering how easy it is to avoid the macro...

**Comment by rgommers** [comment]
> Probably I am daft, because it doesn't :), but at this point it doesn't matter. If PyArrow had not used this macro, but had `type(dtype) is np.dtype` in Python, the situation would be identical! So the "binary" part, I don't get.

It's simply that in Python you only need to deal with one NumPy version - the one used at run-time - while in C you need to deal with two NumPy versions - at build-time and at run-time - and those need to be compatible. As a result, it's harder to fix and more incompatibilities will remain.

**Comment by seberg** [comment]
Do you expect the packager/conda to be able to fix this issue?  That is the only way for me to make head with understanding where you are coming from.  If `PyArrow` had an upper version, then conda would have to wait for `PyArrow` to remove that?
(Which actually makes me wonder if that puts us in the other bad position of less quick testing? `conda` is our best bet at getting our `rc`'s tested quickly!)

> It's simply that in Python you only need to deal with one NumPy version

This doesn't seem true, or at least does not *apply*.  This is absolutely identical, to e.g. deleting some Python alias like `np.int` (assuming that  the project missed a `DeprecationWarning`): in both cases you got incompatible versions, in both cases you update your code to be compatible with all NumPy versions (by not using `np.int`).

The *literally identical* issue exists in Python: the only difference is that you don't have a convenient macro available to help you run into it. But as I mentioned `scipy.ndimage` managed to run into it anyway! Had I not updated SciPy, it would be in the same position as PyArrow without *any* C-code changes necessary (yeah, you would not have to recompile, but compiling is part of every release, so that doesn't matter).

The difference here to "typical deprecations" is that:

* We cannot give a grace period (keeping an old function around), because it is a *switch* we must flip.
* We cannot attach a `DeprecationWarning` or `FutureWarning`.

Neither of those is specific to C (ABI), and both of those are things I have mentioned before many times:
* We could have done a 1.19.0rc1 that would flush out these issues and then back it out for the actual release (we have done this before for relaxed strides, although we did that more than once there), the good old "Oooops release".
* We could have released header-only updates for 1.14.x, 1.15.x (or however far we want to go back).

We did neither, nobody had much appetite for it and now it is literally too late unless we want to skip the 1.20 release entirely (or something almost as drastic).

---

**In any case, for anyone running into this. The fix is to inline the macro and make a new release (or redefine the NumPy one, or...):**
```
#define PyArray_DescrCheck(op) PyObject_TypeCheck(op, &PyArrayDescr_Type)
```
(otherwise be OK with compiling against NumPy 1.16.6+)

Lets talk about what we **CAN** do instead:

* We can delay the actual NumPy 1.20 release until `PyArrow` and `numba` have a chance to do a release, so that anyone running into it can simply update everything.
* We can try to ping more projects directly (but I don't know which ones would be left at this point, so someone would have to try something like using BigQuery)

**Comment by xhochy** [comment]
Looking into a code search on Github, it seems that PyArrow is the only opensource consumer currently. On conda-forge, we can fix all `pyarrow` releases using repodata patches, for new `pyarrow` releases, we can just set a lower bound of 1.16.6. These parts are easily managable. Two things that aren't simple and I'll post them on the Apache Arrow mailing list:

1. We probably won't rebuild old pyarrow wheels (i.e. make no `.post` releases). We did that in the past and it just let to other user frustration. Doing a bugfix release for the older version is also not happening, releases for `pyarrow` are still really expensive on the human effort side and probably nobody is there to take over the load. One of my main intentions* of opening this bug is to have a reference when I close issues in the Apache Arrow JIRA on why things don't work with a newer NumPy. `pyarrow` doesn't really have a stable API, so people don't migrate that quickly as they would do with NumPy.
2. There will most likely be no new Arrow release before the 1.20 NumPy release. It will probably be around the same time but my feeling is that we won't make it in time.

* The other intention was to make people aware that it could also affect other projects. Seems like this is limited to older `numba` versions.

**Comment by seberg** [comment]
Thanks @xhochy please ping if there is anything we can help with.

Again, I am sorry that I missed to ping PyArrow about this a year ago :(, I guess it got lost in the gap between fixing the macro and doing the actual change.  Although I will celebrate if this ends up being one of the more annoying issues :).

**Comment by seberg** [comment]
Closing, pyarrow should be fine now (and about to release).
